gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1360 @@
+
+  static {
+    try {
+      BASE_CONTEXT = new URI(CONTEXT);
+    } catch (URISyntaxException e) {
+      throw new RuntimeException(
+          "Could not initialize LocalDnsHelper due to URISyntaxException.", e);
+    }
+  }
+
+  private long delayChange;
+  private final HttpServer server;
+  private final int port;
+
+  /**
+   * For matching URLs to operations.
+   */
+  private enum CallRegex {
+    CHANGE_CREATE("POST", CONTEXT + "/[^/]+/managedZones/[^/]+/changes"),
+    CHANGE_GET("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/changes/[^/]+"),
+    CHANGE_LIST("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/changes"),
+    ZONE_CREATE("POST", CONTEXT + "/[^/]+/managedZones"),
+    ZONE_DELETE("DELETE", CONTEXT + "/[^/]+/managedZones/[^/]+"),
+    ZONE_GET("GET", CONTEXT + "/[^/]+/managedZones/[^/]+"),
+    ZONE_LIST("GET", CONTEXT + "/[^/]+/managedZones"),
+    PROJECT_GET("GET", CONTEXT + "/[^/]+"),
+    RECORD_LIST("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/rrsets");
+
+    private String method;
+    private String pathRegex;
+
+    CallRegex(String method, String pathRegex) {
+      this.pathRegex = pathRegex;
+      this.method = method;
+    }
+  }
+
+  /**
+   * Wraps DNS data by adding a timestamp and id which is used for paging and listing.
+   */
+  static class RrsetWrapper {
+    static final Function<ResourceRecordSet, RrsetWrapper> WRAP_FUNCTION =
+        new Function<ResourceRecordSet, RrsetWrapper>() {
+          @Override
+          public RrsetWrapper apply(ResourceRecordSet input) {
+            return new RrsetWrapper(input);
+          }
+        };
+    private final ResourceRecordSet rrset;
+    private final Long timestamp = System.currentTimeMillis();
+    private String id;
+
+    RrsetWrapper(ResourceRecordSet rrset) {
+      // The constructor creates a copy in order to prevent side effects.
+      this.rrset = new ResourceRecordSet();
+      this.rrset.setName(rrset.getName());
+      this.rrset.setTtl(rrset.getTtl());
+      this.rrset.setRrdatas(ImmutableList.copyOf(rrset.getRrdatas()));
+      this.rrset.setType(rrset.getType());
+    }
+
+    void setId(String id) {
+      this.id = id;
+    }
+
+    String id() {
+      return id;
+    }
+
+    /**
+     * Equals does not care about the listing id and timestamp metadata, just the rrset.
+     */
+    @Override
+    public boolean equals(Object other) {
+      return (other instanceof RrsetWrapper) && Objects.equals(rrset, ((RrsetWrapper) other).rrset);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(rrset);
+    }
+
+    ResourceRecordSet rrset() {
+      return rrset;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("rrset", rrset)
+          .add("timestamp", timestamp)
+          .add("id", id)
+          .toString();
+    }
+  }
+
+  /**
+   * Associates a project with a collection of ManagedZones. Thread safe.
```

Comment:
I personally don't like calling classes "thread safe" in comments because thread safety also depends on part on the usage.  My preference is to remove "thread safe" from this comment and other ones below.